llvm/lib/SYCLLowerIR/LowerESIMDVecArg.cpp

```diff
@@ -0,0 +1,432 @@
+  Function *NF = Function::Create(NFTy, F.getLinkage(), F.getName());
+  NF->copyAttributesFrom(&F);
+  NF->setCallingConv(F.getCallingConv());
+
+  F.getParent()->getFunctionList().insert(F.getIterator(), NF);
+  NF->takeName(&F);
+  NF->setSubprogram(F.getSubprogram());
+
+  // Now to splice the body of the old function into the new function
+  NF->getBasicBlockList().splice(NF->begin(), F.getBasicBlockList());
+
+  for (unsigned int I = 0; I != F.arg_size(); I++) {
+    auto Arg = F.getArg(I);
+    Type *newTy = argIsSimdPtr(Arg);
+    if (newTy) {
+      // Insert bitcast
+      // bitcast vector* -> simd*
+      auto BitCast = new BitCastInst(NF->getArg(I), Arg->getType());
+      NF->begin()->getInstList().push_front(BitCast);
+      ToModify.insert(std::make_pair(Arg, nullptr));
+      Arg->replaceAllUsesWith(BitCast);
+    }
+  }
+
+  // Loop over the argument list, transferring uses of the old arguments to the
+  // new arguments, also tranferring over the names as well
+  Function::arg_iterator I2 = NF->arg_begin();
+  unsigned int ArgNo = 0;
+  for (Function::arg_iterator I = F.arg_begin(), E = F.arg_end(); I != E;
+       ++I, ++I2, ArgNo++) {
+    auto ArgIt = ToModify.find(I);
+    if (ArgIt == ToModify.end()) {
+      // Transfer old arguments as is
+      I->replaceAllUsesWith(I2);
+      I2->takeName(I);
+    }
+  }
+
+  // Fix call sites
+  SmallVector<std::pair<Instruction *, Instruction *>, 10> OldNewInst;
+  for (auto &use : F.uses()) {
+    // Use must be a call site
+    SmallVector<Value *, 10> Params;
+    auto User = use.getUser();
+    if (isa<CallInst>(User)) {
+      auto Call = cast<CallInst>(User);
+      for (unsigned int I = 0,
+                        NumOpnds = cast<CallInst>(Call)->getNumArgOperands();
+           I != NumOpnds; I++) {
+        auto SrcOpnd = Call->getOperand(I);
+        auto NewTy = argIsSimdPtr(SrcOpnd);
+        if (NewTy) {
+          auto BitCast = new BitCastInst(SrcOpnd, NewTy, "", Call);
+          Params.push_back(BitCast);
+        } else {
+          if (SrcOpnd != &F)
+            Params.push_back(SrcOpnd);
+          else
+            Params.push_back(NF);
+        }
+      }
+
+      // create new call instruction
+      auto NewCallInst = CallInst::Create(NFTy, NF, Params, "");
+      NewCallInst->setCallingConv(F.getCallingConv());
+      OldNewInst.push_back(std::make_pair(Call, NewCallInst));
+    }
+  }
+
+  for (auto InstPair : OldNewInst) {
+    auto OldInst = InstPair.first;
+    auto NewInst = InstPair.second;
+    ReplaceInstWithInst(OldInst, NewInst);
+  }
+
+  F.eraseFromParent();
+
+  return NF;
+}
+
+bool ESIMDLowerVecArgPass::hasGlobalConstExpr(Value *V,
+                                              GlobalVariable *&Global) {
+  if (isa<GlobalVariable>(V)) {
+    Global = cast<GlobalVariable>(V);
+    return true;
+  }
+
+  if (isa<ConstantExpr>(V)) {
+    auto FirstOpnd = cast<ConstantExpr>(V)->getOperand(0);
+    return hasGlobalConstExpr(FirstOpnd, Global);
```

Comment:
Would Value::stripPointerCasts() or one of its variants achieve the same effect without recursion?

Reply:
yes, i tried it and it worked. thanks for the tip. but eventually i did away with this function in latest patch.